packages/itwinui-react/src/core/Table/columns/selectionColumn.tsx

```diff
@@ -81,6 +81,13 @@ export const SelectionColumn = <T extends Record<string, unknown>>(
         title='' // Removes default title that comes from react-table
         disabled={isDisabled?.(row.original)}
         onClick={(e) => e.stopPropagation()} // Prevents triggering on row click
+        onChange={() => {
+          if (row.subRows.length > 0) {
+            row.toggleRowSelected(!row.isSomeSelected);
+          } else {
+            row.toggleRowSelected();
+          }
```

Comment:
I'm not sure I understand this logic, and the PR description feels too vague. Can you try to explain it in more detail?

Also:
1. What exactly do you mean by "similar to the one used on the header row"?
2. Can you add a screen recording and the code you used for testing it in vite-playground?
3. Have you verified whether the [behavior difference in expand vs collapse](https://github.com/iTwin/iTwinUI/issues/1972#issuecomment-2034728282) is fixed?
4. Have you investigated whether #928 is related?

Reply:
The logic here is that the row selection should be toggling to false instead of true if some sub rows are currently selected, since there are situations, like when the row has disabled sub rows, where not all of the sub rows can be selected. If you keep trying to select all sub rows when some sub rows can't be selected, then the checkbox can never toggle back to false. On retesting when adding back the vite-playground code, did notice an issue with using only `isSomeSelected` in the logic, since that value is `false` when all sub rows are selected. Reworking the logic to fix this, and will add a test for it.

1.  The checkbox for the selection column's header row has an `onChange` handler that uses similar logic, and never had these issues to begin with as a result.
2. 

https://github.com/iTwin/iTwinUI/assets/141063311/e5788835-417b-42bb-8c44-2e266d769680

Code has also been readded in latest commit for the vite-playground
3. The odd behavior with the expand and collapse was caused by the strange limbo state that the checkbox was being put in, where it was trying to become checked, but no sub rows were updating and all of the sub rows were not selected, so it would temporarily become unchecked. With the selection state fixed, this issue is also fixed.
4. This update to the state also seems to have fixed that issue. Relevant screen capture: 

https://github.com/iTwin/iTwinUI/assets/141063311/2d03338b-210d-4cf1-900a-581d7a4f3f43

Reply:
> The logic here is that the row selection should be toggling to false instead of true if some sub rows are currently selected, since there are situations, like when the row has disabled sub rows, where not all of the sub rows can be selected. 

A condensed version of this explanation would be a useful thing to add as an in-code comment.

> If you keep trying to select all sub rows when some sub rows can't be selected, then the checkbox can never toggle back to false. 

This makes sense. So it should go between checked vs unchecked, and the indeterminate state should be independent of the checked state, correct?

> On retesting when adding back the vite-playground code, did notice an issue with using only `isSomeSelected` in the logic, since that value is `false` when all sub rows are selected. Reworking the logic to fix this, and will add a test for it.

So is this fixed or are you still working on it?

---

> 1. […] The checkbox for the selection column's header row has an `onChange` handler that uses similar logic, and never had these issues to begin with as a result.

When talking about code on github, you should [permalink it](https://github.com/iTwin/iTwinUI/assets/9084735/f5b79f95-52b1-4012-bdc6-53cfd72f8a66) to make it easier for review.

> 2. […] Code has also been readded in latest commit for the vite-playground

The playground code is useful for the deploy preview, but it should also be referenced in the PR description, because the playground will need to be reverted before merge.

> 4. This update to the state also seems to have fixed that issue.

If that issue is also being tackled, then you should [link it to this PR](https://github.com/iTwin/iTwinUI/assets/9084735/b664bb1c-6d8f-487d-bdb6-dff1c153af93).

However, I noticed some incorrect state when filtering. In the screen recording below, you can see that the checkbox does react to user action, but it goes to fully unchecked, when it should instead stay as indeterminate, because some sub-rows are still selected (even if they are not currently visible due to filtering).

https://github.com/iTwin/iTwinUI/assets/9084735/459445b6-b0c8-4e1f-954d-0d3136caf51d

This might be a difficult one to truly fix though. When some sub-rows are disabled and other sub-rows are filtered out, then we have no way to visually distinguish between these two states; they would both be indeterminate (unless I'm missing something).

Reply:
> A condensed version of this explanation would be a useful thing to add as an in-code comment.

Will add.
> So it should go between checked vs unchecked, and the indeterminate state should be independent of the checked state, correct?

Yes, indeterminate is an independent state from checked and unchecked.
> So is this fixed or are you still working on it?

This is fixed, just wanted to add an explanation for why the check changed to include `row.isSelected`.
> The playground code is useful for the deploy preview, but it should also be referenced in the PR description, because the playground will need to be reverted before merge.

Added comment to PR description about removing vite-playground code.
> If that issue is also being tackled, then you should [link it to this PR](https://github.com/iTwin/iTwinUI/assets/9084735/b664bb1c-6d8f-487d-bdb6-dff1c153af93).

Added link to issue to this PR.

> However, I noticed some incorrect state when filtering.

I'll look into fixing this, since I think that maybe the `initalSubRows` value on row might give us a way to check if filtered sub rows are selected.

Reply:
Fixed the filtered sub row deselection issue and added an e2e test for it with this [commit](https://github.com/iTwin/iTwinUI/pull/2013/commits/f7aead31c47928b1ae10be776b6f02c6a837f3b1).

Reply:
I'm still seeing some strange behavior / #928 is unfixed.

https://github.com/iTwin/iTwinUI/assets/9084735/f7235e02-4200-4d49-bf28-1e53f6addfbc

Reply:
That seems to be the inverse of the original problem, where the parent row was trying to select sub rows that couldn't be selected. Now, it is trying to deselect rows that cannot be deselected. I think the best solution here is to have the selection logic ignore disabled and filtered out rows. I'll try to implement that and see how that works.

Reply:
Updated selection logic to account for this in this [commit](https://github.com/iTwin/iTwinUI/pull/2013/commits/bf35a7379df047dfdf2b8448e07517314739d3f7). Also updated e2e tests and added a comment to the handler explaining the logic.

Reply:
The behavior looks good to me now. You should add a screen-recording (or two) in your PR description, along with a [permalink](https://github.com/iTwin/iTwinUI/blob/ed0b545cef47db320f621f43efbe29a24078e409/playgrounds/vite/src/App.tsx) to the playground code (after which you can revert the playground).

As for the code, I would suggest writing it like this for more clarity.
```jsx
onChange={() => {
  if (row.subRows.length > 0) {
    // We want to ignore any sub-rows that are not available (i.e. disabled or filtered out).
    // If all available sub-rows are selected, then we deselect them all, otherwise we select them all.
    const nextState = !row.subRows.every(
      (subRow) => subRow.isSelected || isDisabled?.(subRow.original),
    );
    row.toggleRowSelected(nextState);
  } else {
    row.toggleRowSelected();
  }
}}
```

Lastly, the tests could be made clearer too. It is easier to understand these tests in terms of nested indexes ("Row 1", "Row 1.2", "Row 1.2.1", etc), rather than vague names (`subRowCheckbox`/`secondRowCheckbox`) and vague selectors (`checkboxes.nth(2)`).

Reply:
Alright, will do.

Reply:
Code updates pushed in the latest batch of commits. Also updated the PR description with the screen recordings and the permalink to the playground code.